web3.js/src/util/send-and-confirm-transaction.ts

```diff
@@ -33,18 +34,36 @@ export async function sendAndConfirmTransaction(
     sendOptions,
   );
 
-  const status = (
-    await connection.confirmTransaction(
-      signature,
-      options && options.commitment,
-    )
-  ).value;
+  const status = await connection.getSignatureStatus(signature);
 
-  if (status.err) {
-    throw new Error(
-      `Transaction ${signature} failed (${JSON.stringify(status)})`,
+  const checkBlockHeight = async () => {
+    const blockHeight = await connection.getBlockHeight(
+      options && options.commitment,
     );
-  }
 
+    if (blockHeight > transaction.lastValidBlockHeight!) {
+      throw new Error('Transaction has expired.');
+    } else {
+      if (status) {
```

Comment:
Unless I've missed it, I don't think we ever ask for the signature status again here, right? This just recurses until the block height is exceeded, whether or not the signature ever gets confirmed?